### PR TITLE
fix: `onProgramAccountChange()` and `onAccountChange()` now accept an encoding

### LINKED ISSUE #2725: WSS method onAccountChange doesn't support custom encoding
## Motivation
The Solana RPC method `accountSubscribe` does provide a way to pass encoding so should the JS SDK

## Example use case
```ts
solanaConnection.onAccountChange(
  publicKey,
  (updatedProgramInfo, context) =>
    console.log(updatedProgramInfo.data),
  "confirmed"
);
```

The way this function has the commitment arguement it should have one for encoding too. Maybe just make the third arguement take an object with commitment and encoding as two fields of it.

Right now the data the above code logs is like -

```
<Buffer 06 00 00 00 00 00 00 00 fe 00 00 00 00 00 00 00 07 00 00 00 00 00 00 00 03 00 00 00 00 00 00 00 09 00 00 00 00 00 00 00 09 00 00 00 00 00 00 00 01 00 ... 702 more bytes>
```

This data can't be much used and JSON data will be much more appreciated like the `accountSubscribe` method returns.

### PR DESCRIPTION
**Pull request by steveluscher**
Fixes #2725.

### CODE REVIEW

**Comment by steveluscher** [comment]
* **#2861** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2861?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @steveluscher and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/library-legacy/src/connection.ts [hunk 1]
```diff
@@ -6337,15 +6370,31 @@ export class Connection {
...
+    callback: AccountChangeCallback,
+    config: AccountSubscriptionConfig,
+  ): ClientSubscriptionId;
+  /** @deprecated Instead, pass in an {@link AccountSubscriptionConfig} */
+  // eslint-disable-next-line no-dupe-class-members
   onAccountChange(
     publicKey: PublicKey,
     callback: AccountChangeCallback,
     commitment?: Commitment,
+  ): ClientSubscriptionId;
```
Comment (steveluscher) [inline packages/library-legacy/src/connection.ts:6384]: Existing method signature for backward compatibility.

File: packages/library-legacy/src/connection.ts [hunk 2]
```diff
@@ -6398,17 +6447,37 @@ export class Connection {
...
+    config: ProgramAccountSubscriptionConfig,
+  ): ClientSubscriptionId;
+  /** @deprecated Instead, pass in a {@link ProgramAccountSubscriptionConfig} */
+  // eslint-disable-next-line no-dupe-class-members
   onProgramAccountChange(
     programId: PublicKey,
     callback: ProgramAccountChangeCallback,
     commitment?: Commitment,
     filters?: GetProgramAccountsFilter[],
+  ): ClientSubscriptionId;
```
Comment (steveluscher) [inline packages/library-legacy/src/connection.ts:6462]: Existing method signature for backward compatibility.

File: packages/library-legacy/src/connection.ts [hunk 3]
```diff
@@ -6398,17 +6447,37 @@ export class Connection {
...
     const args = this._buildArgs(
       [programId.toBase58()],
       commitment || this._commitment || 'finalized', // Apply connection/server default.
       'base64' /* encoding */,
-      filters ? {filters: filters} : undefined /* extra */,
+      config
+        ? config
+        : maybeFilters
+          ? {filters: maybeFilters}
+          : undefined /* extra */,
```
Comment (steveluscher) [inline packages/library-legacy/src/connection.ts:6480]: According to TypeScript, you either have a config, or you have commitment/filters, but never both.

File: packages/library-legacy/src/connection.ts [hunk 4]
```diff
@@ -2787,6 +2787,39 @@ export type GetNonceAndContextConfig = {
   minContextSlot?: number;
 };
 
+export type AccountSubscriptionConfig = Readonly<{
+  /** Optional commitment level */
+  commitment?: Commitment;
+  /**
+   * Encoding format for Account data
+   *   - `base58` is slow.
```
Comment (buffalojoec) [inline packages/library-legacy/src/connection.ts:2795]: I laughed at this, until I realized it's actually what's in the docs. 😅

File: packages/library-legacy/src/connection.ts (hunk 3 above)
Comment (buffalojoec) [reply packages/library-legacy/src/connection.ts:6479]: Little gross, but wfm.

**Review by buffalojoec — APPROVED** [review]
Looks great!

**Pull request merged**